[FEAT] Base user roles

Liinkiing opened the pull request:
## Related issue(s)
<!--
Link your pull request to an issue using a keyword:
https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->
#25 


## Description
<!-- Describe globally your PR in its context -->
This PR enable the ability to create and edit a user in the application. It also add the support of a roles field in the user to check it's privileges inside the application. These roles are used mostly to restrict access / actions to some parts of the application. It is not the permissions based on a given account, this is something else that deserve it's own PR.

I've also added the support to use local cloud functions URL in the app, you juste have to provide a env var `VUE_APP_LOCAL_FIREBASE_CF_URL`.

I've also done some changes to the login workflow, I've tested it and it seems that I have not broken anything, but double check the code in case. It allows us to have proper support for the token claims, it means that when an administrator change a user role, it is reflected in REALTIME and a new token is issued based on his new claims. So it means that if I am connected as a SuperAdmin, someone decide to change my role to User, my UI will adapt as soon as the person made the change, and we no longer need to refresh / logout the user to have an updated token!

I changed how we store the logged in user in the Vuex store. Previously, we stored directly the user object from Firebase which was in fact a `UserRecord` and had a tooooons of methods and properties (it was not a plain object but looked more like a class or something like that). So we instead store the JSON representation of that user in the Vuex. Before that, when inspecting the store and showing the user in the devtool, it crashed the page due to how large the object was and maybe it had circular dependencies.

**Note**: for the moment, only me, Erwann and Romain account have the SuperAdmin role, so only those account can access to this page

## What does this PR do?
<!-- List all stuff that have been done -->
- [x] Add a role system for the users
- [x] Add a way to use local cloud functions URL
- [x] Expose some constants and middleware in the package
- [x] Add Create and Edit user modal dialog and behaviour
- [x] Add a global `$isGranted` method, which allows us to easily check if the logged in user has some permissions, to allows for example to conditionally hide some UI elements based on it's role
- [x] Store JSON representation of the firebase user in the Vuex store

## Deploy notes
<!-- Add additional instructions is necessary to deploy this PR -->
Review & merge.


## Steps to test or reproduce
<!-- List different steps to approve your PR -->
1. Connect the user
2. Change some roles in the /users page


## Impacted area(s) in Application
<!-- List impacted areas by being generally -->
- Users listing
- User login

Eronne commented on src/router/middleware/user/super-admin.ts:
```diff
@@ -0,0 +1,17 @@
+  nextPipeline: ReturnType<MiddlewarePipeline>
+) => {
+  if (
+    userStore.state.user &&
+    userStore.state.user.role &&
+    userStore.state.user.role.code >= Roles.SuperAdmin
```
Not possible with optional chaining?

Liinkiing replied:
Yep, totally, will make the changes 👍

Eronne commented on src/store/user/index.ts:
```diff
@@ -3,6 +3,8 @@ import { IVueI18n } from "vue-i18n";
 import firebase from "firebase";
 import { LS_LANGUAGE_KEY } from "@/init/plugins/vue-i18n";
 import { Collections } from "@/constants/firebase";
+import * as fb from "firebase";
```
Not used import

Liinkiing commented:
> Juste leaved some comments.
> 
> I cannot review `.vue` file behavior & app globally because after running `yarn serve` I am facing two issues :
> 
> 1. Typing error (not major) : 
> ```bash
> ERROR in /.../maestro/src/init/router.ts(18,48):
> 18:48 Type '{ state: { user: {}; isAuthenticated: boolean; }; getters: { user: (state: State) => User; isGranted: (state: State) => (role: Roles) => boolean; isAuthenticated: (state: State) => boolean; }; mutations: { ...; }; actions: { ...; }; }' is missing the following properties from type 'Store<any>': replaceState, dispatch, commit, subscribe, and 6 more.
>     16 |     if (!to.meta.middleware) return next();
>     17 |
>   > 18 |     const context: Context = { to, from, next, userStore };
>        |                                                ^
>     19 |     const middleware: Middleware[] = to.meta.middleware;
>     20 |     const index = 0;
>     21 |
> ERROR in /.../maestro/src/store/user/index.ts(6,1):
> 6:1 'fb' is declared but its value is never read.
>     4 | import { LS_LANGUAGE_KEY } from "@/init/plugins/vue-i18n";
>     5 | import { Collections } from "@/constants/firebase";
>   > 6 | import * as fb from "firebase";
>       | ^
>     7 | import { Roles } from "@/constants";
>     8 |
>     9 | type State = {
> ```
> 
> 2. App cannot load (major) :
> `Uncaught ReferenceError: regeneratorRuntime is not defined`
> 
> After some research, I saw that it is probably due to async use in `src/init/index.ts` 
> (line 72) `firebase.auth().onAuthStateChanged(async (user: fb.User) => {...}`
> 
> ![image](https://user-images.githubusercontent.com/10868310/83511140-77e5f980-a4ce-11ea-920a-1dc2013d3d0e.png)
> 

What cannot load? The application product autopilot or the dev folder in maestro?

Liinkiing commented on src/cloud-functions/permissions.ts:
```diff
@@ -0,0 +1,9 @@
+  return context.auth?.token?.role >= role;
+};
+
+export const isSuperAdmin = (context: any) => {
+  return context.auth?.token?.role >= Roles.SuperAdmin;
+};
```
No because in the vue app the check comme from the vuex store, and in the cloud function it comes from the context Auth if the request, I cannot use the same code for the two

Liinkiing commented on src/cloud-functions/users/create-user.ts:
```diff
@@ -0,0 +1,36 @@
+          role: data.role
+        })
+        .then(() => {
+          const metadataRef = admin
+            .database()
+            .ref("metadata/" + result.data.uid);
+          return metadataRef.set({ refreshTime: new Date().getTime() });
```
Exact 👍

Liinkiing commented on src/cloud-functions/users/update-user.ts:
```diff
@@ -0,0 +1,50 @@
+): Partial<UpdateUserPayload> => {
+  let result = {
+    email: payload.email,
+    emailVerified: false,
+    password: payload.password,
+    displayName: payload.displayName,
+    photoURL: payload.photoURL,
+    disabled: false
```
For this behavior I've did like in Jarvis studio. So you want me to set the email verified to true and not setting at all the disabled property?

Eronne replied:
No I am just saying that for user creation, it is justified to set `emailVerified` to false because it is a fresh user.

But while updating user properties, if he already verified his email address (`emailVerified` is `true`), with this update, it will be reseted to false.


And for the disable property, we have to use `payload.disabled` to set a property from the client with this cloud function or use a dedicated cloud function, what do you think ?

Liinkiing replied:
Using `payload.disabled` looks good to me 👍

Liinkiing commented:
@Eronne I've fixed the error with babel and async / await, but I think we should not test directly the `dev` folder, but instead checkout the branch with the same name in product autopilot and consume product autopilot. It will reflect more the real use case, and more impotant it will avoid to duplicate the code. For exemple, you cannot test in the dev app the addition / edition of user because I would have to also create the cloud functions in the maestro playground. 

When we will have more features, it will be hard to maintain both the dev app within maestro and keep them in sync with features we add. 

But I've fixed the problem with regeneratorRuntime and the dev app within maestro should work correctly now 👍

Eronne commented:
I am just facing some troubles to load both applications (`maestro-dev` & a consumer application).

Loader is infinite.

Eronne commented on src/cloud-functions/users/update-user.ts:
```diff
@@ -0,0 +1,50 @@
+  return result;
+};
+
+export const updateUser = (admin: any, data: UpdateUserPayload, _: any) => {
+  return admin
+    .auth()
+    .updateUser(data.uid, cleanUpdatePayload(data))
+    .then((user: any) => {
+      return { data: user };
+    })
+    .then((result: any) => {
+      const userRole = ROLES[(data.role ?? 0) as Roles];
+      console.log(
+        `Setting user ${result.data.email} to role ${userRole.name} (${userRole.code})`
+      );
+      return admin
+        .auth()
+        .setCustomUserClaims(result.data.uid, {
+          role: data.role
+        })
+        .then(() => {
+          const metadataRef = admin
+            .database()
+            .ref("metadata/" + result.data.uid);
+          return metadataRef.set({ refreshTime: new Date().getTime() });
+        });
+    })
+    .catch((error: Error) => error);
+};
```
While updating a user, we also have to update its "social" information into users collection

Eronne commented on src/components/dialogs/CreateUserDialog.vue:
```diff
@@ -0,0 +1,157 @@
+<template>
+  <el-dialog
```
Cannot close dialog while clicking on cross.

Eronne commented on src/components/dialogs/UpdateUserDialog.vue:
```diff
@@ -0,0 +1,148 @@
+<template>
+  <el-dialog
```
Cannot close dialog while clicking on cross.
